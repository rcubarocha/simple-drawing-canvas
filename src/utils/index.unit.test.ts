import { getCanvasCoordsFromEvent, drawLine } from '.';

describe('get canvas-space coords from mouse event and controller config', () => {
  const c = document.createElement('canvas');
  c.width = 300;
  c.height = 300;

  c.getBoundingClientRect = jest.fn(() => ({
    x: 50,
    y: 50,
    width: 100,
    height: 100,
    top: 50,
    right: 150,
    bottom: 150,
    left: 50,
  } as DOMRect));

  test('in-bounds events should be converted correctly', () => {
  
    let me = new window.MouseEvent('mousedown', {
      clientX: 100,
      clientY: 100,
    });
  
    const center = getCanvasCoordsFromEvent(me, c, {
      scale: 3, width: 300, height: 300, background: null,
    });
  
    expect(center).toEqual({ x: 150, y: 150 });
  
    me = new window.MouseEvent('mousedown', {
      clientX: 50,
      clientY: 50,
    });
  
    const topLeft = getCanvasCoordsFromEvent(me, c, {
      scale: 3, width: 300, height: 300, background: null,
    });
  
    expect(topLeft).toEqual({ x: 0, y: 0 });
  
    me = new window.MouseEvent('mousedown', {
      clientX: 150,
      clientY: 50,
    });
  
    const topRight = getCanvasCoordsFromEvent(me, c, {
      scale: 3, width: 300, height: 300, background: null,
    });
  
    expect(topRight).toEqual({ x: 300, y: 0 });
  
    me = new window.MouseEvent('mousedown', {
      clientX: 50,
      clientY: 150,
    });
  
    const bottomLeft = getCanvasCoordsFromEvent(me, c, {
      scale: 3, width: 300, height: 300, background: null,
    });
  
    expect(bottomLeft).toEqual({ x: 0, y: 300 });
  
    me = new window.MouseEvent('mousedown', {
      clientX: 150,
      clientY: 150,
    });
  
    const bottomRight = getCanvasCoordsFromEvent(me, c, {
      scale: 3, width: 300, height: 300, background: null,
    });
  
    expect(bottomRight).toEqual({ x: 300, y: 300 });
  });

  test('out-of-bounds events should be converted correctly', () => {
  
    const me = new window.MouseEvent('mousedown', {
      clientX: 20,
      clientY: 245,
    });
  
    const outOfBounds = getCanvasCoordsFromEvent(me, c, {
      scale: 3, width: 300, height: 300, background: null,
    });
  
    expect(outOfBounds).toEqual({ x: -90, y: 585 });
  });

});

describe('drawLine should accurately draw on canvas', () => {
  let canvas: HTMLCanvasElement;
  let ctx: CanvasRenderingContext2D;

  beforeEach(() => {
    canvas = document.createElement('canvas');
    canvas.width = 300;
    canvas.height = 300;

    ctx = canvas.getContext('2d')!;
  });

  test('should match the same triangle snapshot', () => {
    drawLine(ctx, { x: 50, y: 250 }, { x: 150, y: 50 }, 10, '#000000');
    drawLine(ctx, { x: 150, y: 50 }, { x: 250, y: 250 }, 10, '#000000');
    drawLine(ctx, { x: 250, y: 250 }, { x: 50, y: 250 }, 10, '#000000');

    expect(canvas.toDataURL()).toMatchSnapshot()
  });
});