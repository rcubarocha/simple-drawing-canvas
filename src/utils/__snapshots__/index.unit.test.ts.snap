// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawLine should accurately draw on canvas should match the same triangle snapshot 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAYAAAB5fY51AAAABmJLR0QA/wD/AP+gvaeTAAAGq0lEQVR4nO3aPXJdRRAG0EYkFBkEsAsWAAtgCSJhMY4ICMmJYQ+kZgGUvQYSlyBwbBFIwrbsp/fuvfPXM+dUKaEc3Krp/r62TAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMIVnEXETEW/uf27u/xvAUJ5HxO2Jn+cdvwvgPc/idFg9/Li0gO6uIuJ1nA+s1/d/FqCb6zgfVg8/152+ESCuIuKvuDywXoYrC+hky3XlygK6uYqIF7E9sF6EKwtobM915coCmtv6uyu/ywK6OXJdubKAZo5eV64soJkS15UrC6iu1HXlygKqK3ldubKAakpfV64soJot19XL2BZuriygmK3X1XVsDzhXFlDEnvDZE3IAhxwJHlcW0NSR0HFlAc2UCBxXFtBEibBxZQHVlQwaVxZQVcmQcWUB1dQIGFcWUEWNcHFlAcXVDBZXFlBUzVBxZQHFtAgUVxZQRIswcWUBh7UMElcWcEjLEHFlAbv1CBBXFrBLj/BwZQGb9QwOVxawSc/QcGUBFxshMFxZwEVGCIsRQhMY3EhBMUJwAgMbKSRGCk9gMCMGxEgBCgxkxHAYMUSBzkYOhhGDFOho5FAYOUyBxjIEwsiBCjSUIQwyhCpQWaYgyBCsQEWZQiBTuAKFZQyATAELFJRx+TOGLHBQ5sXPGLTAAZmXPnPYAhvNsPCZAxfYYIZlnyF0gTNmWvQZghd4wkxLPlP4Ao/MuOAzBTDwjhmXe8YQhuXNvNgzBjEsbealnjmMYTkrLPTMgQxLWWGZVwhlmN5Ki7xCMMPUVlrilcIZprPiAq8U0DCVFZd3xZCG9FZe3BWDGlJbeWlXDmtIx8KuHdiQimUV2pCCRX1LcMPgLOlbwhsGZkE/JMBhUJbzQ0IcBmQxTxPkMBhLeZowh4FYyPMEOgzCMp4n1GEAFvFygh06s4SXE+7QkQXcTsBDJ5ZvOyEPHVi8/QQ9NGbp9hP20JCFO07gQyOW7TihDw1YtHIEP1RmycoR/lCRBStPAUAllqs8JQAVWKx6FAEUZqnqUQZQkIWqTyFAIZapPqUABVikdhQDHGSJ2lEOcIAFak9BwE6Wpz0lATtYnH4UBWxkafpRFrCBhelPYcCFLEt/SgMuYFHGoTjgDEsyDuUBT7Ag41EgcILlGI8SgY+wGONSJPCIpRiXMoF3WIjxKRS4ZxnGp1QgLEImioXlWYI8lAtLswD5KBiWZfjzUTIsyeDnpWhYjqHPS9mwFAOfn8JhGYY9P6XDEgz6PBQP0zPk81A+TM2Az0cBMS3DPR8lxJQM9rwUEdMx1PNSRkzFQM9PITENwzw/pcQUDPI6FBPpGeJ1KCdSM8DrUVCkZXjXo6RIyeCuS1GRjqFdl7IiFQOLwiINw4rSIgWDygPFxfAMKQ+UF0MzoDymwBiW4eQxJcaQDCanKDKGYyg5RZkxFAPJOQqNYRhGzlFqDMEgcinFRneGkEspN7oygGyl4OjG8LGVkqMLg8deio7mDB17KTuaMnAcpfBoxrBxlNKjCYNGKYqP6gwZpSg/qjJglKYAqcZwUZoSpAqDRS2KkOIMFbUoQ4oyUNSmECnGMFGbUqQIg0QripHDDBGtKEcOMUC0piDZzfDQmpJkF4NDL4qSzQwNvShLNjEw9KYwuZhhoTelyUUMCqNQnJxlSBiF8uRJBoTRKFBOMhyMRonyUQaDUSlSPmAoGJUy5T0GgtEpVP5nGBidUiUiDAJ5KFYMAWko18UZALJRsAvz+GSjZBfl4clK0S7Io5OVsl2MByc7hbsQj012SncRHppZKN4FeGRmoXwn54GZjQKemMdlNkp4Uh6WWSniCXlUZqWMJ+NBmZ1CnojHZHZKeRIeklUo5gl4RFahnJPzgKxGQSfm8ViNkk7Kw7EqRZ2QR2NVyjoZD8bqFHYiHovVKe0kPBTcWaa4P+n9AQdcR8RvvT8CEvohIn7v/RF7ZA2sq4j4OyK+6v0hkNCriPg6It70/pCtPu39ATv9FBHf9/4ISOrziPgsIv7o/SFbZb2wbiLii94fAYn9ExFf9v6IrbIG1pvI++0wgttI+Mv3dB8MrCtrYP3b+wMguZQ7lDWwfun9AZBcyh3K/Hug5xHxbe+PgIT+jIjven/EHln/t4aIiF/j7kL8Ju7+iTZz+EJtt3H318CfI+LHzt8CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQEb/AfFq9oXssdzyAAAAAElFTkSuQmCC"`;
